src/publish-build/index.ts

```diff
@@ -0,0 +1,90 @@
+import * as core from '@actions/core';
+import {
+  S3Client,
+  PutObjectCommand,
+  PutObjectCommandOutput,
+  PutObjectCommandInput,
+} from '@aws-sdk/client-s3';
+import { promises as fs, createReadStream } from 'fs';
+import mime from 'mime';
+import * as path from 'path';
+
+const REQUIRED_ENV_VARS = [
+  'AWS_ACCESS_KEY_ID',
+  'AWS_SECRET_ACCESS_KEY',
+  'AWS_BUCKET',
+  'PROJECT',
+  'SOURCE_DIR',
+  'DESTINATION_DIR',
+  'COMMIT_HASH',
+];
+
+function validateEnvVars(): void {
+  for (const requiredVar of REQUIRED_ENV_VARS) {
+    if (!process.env[requiredVar]) {
+      const errMsg = `Required env variable missing: ${requiredVar}`;
+      core.error(errMsg);
+      throw new Error(errMsg);
+    }
+  }
+}
+
+async function getFileNames(dir: string): Promise<string[]> {
+  const dirents = await fs.readdir(dir, { withFileTypes: true });
+  const files = await Promise.all(
+    dirents.map(async dirent => {
+      const res = path.resolve(dir, dirent.name);
+      return dirent.isDirectory() ? getFileNames(res) : res;
+    }),
+  );
+  return Array.prototype.concat(...files);
+}
+
+async function upload(
+  sourcePath: string,
+  directoryPath: string,
+  fileNames: string[],
+): Promise<PutObjectCommandOutput[]> {
+  const accessKeyId = process.env['AWS_ACCESS_KEY_ID'] ?? '';
+  const secretAccessKey = process.env['AWS_SECRET_ACCESS_KEY'] ?? '';
+  const project = process.env['PROJECT'] ?? '';
+  const commitHash = process.env['COMMIT_HASH'] ?? '';
+  const destinationDir = process.env['DESTINATION_DIR'] ?? '';
+  const client = new S3Client({
+    region: 'us-east-2',
+    credentials: {
+      accessKeyId,
+      secretAccessKey,
+    },
+  });
+  const bucket = process.env['AWS_BUCKET'] ?? '';
+
+  const uploads = fileNames.map(async fileName => {
+    const key = path.normalize(
+      `${destinationDir}/${project}/${commitHash}/${path.relative(
+        path.normalize(`${sourcePath}/${directoryPath}`),
+        fileName,
+      )}`,
+    );
+    const input: PutObjectCommandInput = {
+      Body: createReadStream(fileName),
+      Key: key,
+      Bucket: bucket,
+      ContentType: mime.getType(path.extname(fileName)) ?? '',
+      CacheControl: 'no-cache',
+    };
+    const command = new PutObjectCommand(input);
```

Comment:
This is something you can definitely take or leave, but AWS has a convenient [class for S3 uploads](https://github.com/aws/aws-sdk-js-v3/blob/main/lib/lib-storage/README.md) that perform pretty efficient uploads. It can break up an individual file upload into chunks to upload concurrently.

[Here's an example from the Autobuilder of it being used](https://github.com/mongodb/docs-worker-pool/blob/0689a7c3520c8f6121373d9f964bfeb9f54ee330/src/cache-updater/index.ts#L59).

But, if this already works fine, and adding this is too much of a hassle, feel free to keep what you have 👍

Reply:
glad to do this in a follow up PR involving file read streams!